### PR TITLE
#7954 alias analysis attempt

### PR DESCRIPTION
**Pull request by badumbatish**
Mentioned in #7954, attempted in #8535, reattempted in this pr.

Draft first for ci pipeline. Possible rebase for better commit msg

### CODE REVIEW

File: cranelift/codegen/src/dominator_tree.rs [hunk 1]
```diff
@@ -439,6 +439,13 @@ impl DominatorTreePreorder {
             }
         }
     }
+
+    /// Clear the data structure
+    /// Recursively clears its underlying children
+    pub fn clear(&mut self) {
+        self.nodes.clear();
+        self.stack.clear();
+    }
```
Comment (elliottt) [inline cranelift/codegen/src/dominator_tree.rs:448]: Is this function still needed? It doesn't look like it's referenced from the changes in this PR.

Comment (badumbatish) [reply cranelift/codegen/src/dominator_tree.rs]: hi there :) I double checked and it seems nothing is calling this function. I've removed it. Thank you for the review :)

File: cranelift/codegen/src/dominator_tree.rs [hunk 2]
```diff
@@ -482,7 +482,42 @@ impl DominatorTreePreorder {
     pub fn dominates(&self, a: Block, b: Block) -> bool {
         let na = &self.nodes[a];
         let nb = &self.nodes[b];
-        na.pre_number <= nb.pre_number && na.pre_max >= nb.pre_max
+        // debug_assert!(na.pre_number != 0);
+        // debug_assert!(nb.pre_number != 0);
```
Comment (elliottt) [inline cranelift/codegen/src/dominator_tree.rs:486]: Does the test suite pass with these assertions uncommented? You're testing `nb` explicitly for being unreachable on line 491, so I'm guessing that at least that one doesn't pass?

File: cranelift/codegen/src/dominator_tree.rs [hunk 3]
```diff
@@ -482,7 +482,42 @@ impl DominatorTreePreorder {
...
+    /// Checks if one instruction dominates another.
+    ///
+    /// From @jameysharp:
+    /// Instruction a dominates instruction b if either of these conditions are true:
+    ///    a and b are in the same block, and a is not after b.
+    ///    Or the block containing a dominates the block containing b.
+    ///
+    ///    "I would prefer to have dominates_inst next to the dominates method that it uses.
+    ///     Then someone looking at one of them can easily see the other one too."
+    ///
```
Comment (elliottt) [inline cranelift/codegen/src/dominator_tree.rs:509]: Could you rephrase this comment from Jamey as a comment about the function implementation?

File: cranelift/codegen/src/dominator_tree.rs (hunk 2 above)
Comment (badumbatish) [reply cranelift/codegen/src/dominator_tree.rs:486]: hi Elliot, yes one of the test suite failed with this assertion.  Here is the summary of the test suite through local ci: 
```
failures:

---- dominator_tree::tests::unreachable_node stdout ----
thread 'dominator_tree::tests::unreachable_node' panicked at cranelift/codegen/src/dominator_tree.rs:486:9:
assertion failed: nb.pre_number != 0


failures:
    dominator_tree::tests::unreachable_node

test result: FAILED. 187 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.21s

error: test failed, to rerun pass `-p cranelift-codegen --lib`
```

File: cranelift/codegen/src/dominator_tree.rs (hunk 3 above)
Comment (badumbatish) [reply cranelift/codegen/src/dominator_tree.rs]: resolved

**Comment by adamperlin** [comment]
Hi @badumbatish! I am wondering what the status of this PR is? I'd potentially be interesting in continuing to work on this issue!

**Comment by badumbatish** [comment]
Ah please take on this issue, I should have closed it since I'm a bit busy with some other work @adamperlin
